Prod 325 databricks rollovers

rmoneys opened the pull request:
https://synccomputing.atlassian.net/browse/PROD-325

Start with spark_log_parser/\_\_main\_\_.py to see how the new event log module is used.

gorskysd commented on spark_log_parser/__main__.py:
```diff
@@ -1,37 +1,42 @@
-    os.remove(result_path)
+if args.log_file.suffixes:
+    result_path = os.path.join(
+        args.result_dir, "parsed-" + args.log_file.name[: -len("".join(args.log_file.suffixes))]
```
Clever way of stripping suffixes!

rmoneys replied:
Yeah, kinda wish there was something in pathlib for this

gorskysd commented on spark_log_parser/extractor.py:
```diff
@@ -0,0 +1,213 @@
+
+        return sub_paths
+
+    FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
```
Might move this to the top of the class for clarity.

rmoneys replied:
In general, I try to keep variables close to the code that uses them, but I think you're right here - it's more helpful and conventional at the top of the class. Thanks!

gorskysd commented on spark_log_parser/extractor.py:
```diff
@@ -0,0 +1,213 @@
+
+    def extract(self) -> list[Path]:
+        if self.source_url.scheme in {"https", "s3"}:
+            local_path = self.work_dir.joinpath(self.source_url.path.split("/")[-1])
```
Alternative to method to get last field without hardcoding "/" character here: https://stackoverflow.com/questions/3925096/how-to-get-only-the-last-part-of-a-path-in-python

rmoneys replied:
`os.path` would work because we'll always be running on a 'nix server. But I don't like to use it on URL paths because it's intended for filesystem paths.

gorskysd commented on spark_log_parser/extractor.py:
```diff
@@ -0,0 +1,213 @@
+
+        return False
+
+    def _add_to_stats_and_verify(self, size, count=1):
```
I like these methods for tracking the starts, I'm going to pirate this for the spark_predictor!

gorskysd commented on spark_log_parser/extractor.py:
```diff
@@ -0,0 +1,213 @@
+        self.file_total -= count
+
+
+    def _download(self, path):
```
The only shortcoming I see here is that it requires the s3 to be a single prefix to an archive or single eventlog. It could be expanded to ingesting a prefix, e.g. using the method I implemented here. https://github.com/synccomputingcode/sync_backend/blob/bbb8c286d03b13668b87ff6f68b5aaaace071598/spark_predictor/internal_tools/analysis/file_copy_handler.py#L46

This isn't necessary, IMO, as long as the README makes clear that s3 links must be a full object path and not a prefix.

gorskysd reviewed (approved):
Nice. The compression handling here is really clear, I like the design pattern you put in. Also the robust use of Path was nice, there are a lot of pieces I saw here that I wasn't familiar with.

I made some minor suggestions for change, but nothing critical IMO, so submitting my approval.

gorskysd reviewed (changes requested):
One last minute change, sorry!

Corner case to consider where there's only one rollover log and it's not the 0th -- in this case I believe the rollover ID check is bypassed. This is a common case when client's don't know to expect more than one eventlog. See example log in https://synccomputing.atlassian.net/browse/PROD-399?focusedCommentId=11208

rmoneys commented:
Phew!
Rolled in some good feedback since [e5f1974](/synccomputingcode/spark_log_parser/compare/e5f1974..PROD-325-databricks-rollovers) including,
1. Added support for s3 directories
2. Handled missing first db rollover log corner case
3. Removed pydantic model since it wasn't doing much besides getting in the way of injecting an s3 client to make testing easier
4. Misc refactor for more robust code 🤞

gorskysd reviewed (approved):
Nice, this looks great, Scott.

I'm learning a lot about Path and url handling here that's super useful.
